sdmetrics/single_table/privacy/disclosure_protection.py

```diff
@@ -0,0 +1,241 @@
+}
+
+
+class DisclosureProtection(SingleTableMetric):
+    """The DisclosureProtection metric."""
+
+    goal = Goal.MAXIMIZE
+    min_value = 0
+    max_value = 1
+
+    @classmethod
+    def _validate_inputs(
+        cls,
+        real_data,
+        synthetic_data,
+        known_column_names,
+        sensitive_column_names,
+        computation_method,
+        continuous_column_names,
+        num_discrete_bins,
+    ):
+        if not isinstance(real_data, pd.DataFrame) or not isinstance(real_data, pd.DataFrame):
+            raise ValueError('Real and synthetic data must be pandas DataFrames.')
+
+        if len(known_column_names) == 0:
+            raise ValueError('Must provide at least 1 known column name.')
+        elif not set(real_data.columns).issuperset(set(known_column_names)):
+            missing = "', '".join(set(known_column_names) - set(real_data.columns))
+            raise ValueError(f"Known column(s) '{missing}' are missing from the real data.")
+
+        if len(sensitive_column_names) == 0:
+            raise ValueError('Must provide at least 1 sensitive column name.')
+        elif not set(real_data.columns).issuperset(set(sensitive_column_names)):
+            missing = "', '".join(set(sensitive_column_names) - set(real_data.columns))
+            raise ValueError(f"Sensitive column(s) '{missing}' are missing from the real data.")
+
+        if computation_method.upper() not in CAP_METHODS.keys():
+            raise ValueError(
+                f"Unknown computation method '{computation_method}'. "
+                f"Please use one of 'cap', 'zero_cap', or 'generalized_cap'."
+            )
+
+        if continuous_column_names is not None and not set(real_data.columns).issuperset(
+            set(continuous_column_names)
+        ):
+            missing = "', '".join(set(continuous_column_names) - set(real_data.columns))
+            raise ValueError(f"Continous column(s) '{missing}' are missing from the real data.")
+
+        if not isinstance(num_discrete_bins, int) or num_discrete_bins <= 0:
+            raise ValueError('`num_discrete_bins` must be an integer greater than zero.')
+
+        super()._validate_inputs(real_data, synthetic_data)
+
+    @classmethod
+    def _get_null_categories(cls, real_data, synthetic_data, columns):
+        base_null_value = '__NULL_VALUE__'
+        null_category_map = {}
+        for col in columns:
+            null_value = base_null_value
+            categories = set(real_data[col].unique()).union(set(synthetic_data[col].unique()))
+            while null_value in categories:
+                null_value += '_'
+
+            null_category_map[col] = null_value
+
+        return null_category_map
+
+    @classmethod
+    def _discretize_column(cls, real_column, synthetic_column, num_bins):
+        bin_labels = [str(x) for x in range(num_bins)]
+        real_binned, bins = pd.cut(real_column, num_bins, labels=bin_labels, retbins=True)
```

Comment:
What happens if the real_column contains NaNs? It seems to work
```python
import pandas as pd
import numpy as np
real_binned, bins = pd.cut(pd.Series([1, 7, 5, 4, 6, pd.NA], dtype='Int64'), 3, retbins=True)
real_binned
bins
```

```python
In [17]: real_binned
Out[17]:
0    (0.994, 3.0]
1      (5.0, 7.0]
2      (3.0, 5.0]
3      (3.0, 5.0]
4      (5.0, 7.0]
5             NaN
dtype: category
Categories (3, interval[float64, right]): [(0.994, 3.0] < (3.0, 5.0] < (5.0, 7.0]]

In [18]: bins
Out[18]: array([0.994, 3.   , 5.   , 7.   ])
```

Reply:
Yeah this is expected, we handle all null values separately after we discretize all the columns.

Reply:
Gotcha, sounds good